polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp

```diff
@@ -1107,27 +1093,59 @@ static bool canBeHoisted(LICMCandidate &candidate, LoopLikeOpInterface loop,
   return true;
 }
 
-// Populate \p LICMCandidates with operations that can be hoisted out of the
-// given loop \p loop.
+/// Populate \p LICMCandidates with operations that can be hoisted out of the
+/// given loop \p loop.
 static void
 collectHoistableOperations(LoopLikeOpInterface loop,
                            const AliasAnalysis &aliasAnalysis,
                            const DominanceInfo &domInfo,
                            SmallVectorImpl<LICMCandidate> &LICMCandidates) {
+  assert(LICMCandidates.empty() && "Expecting empty LICMCandidates");
   // Do not use walk here, as we do not want to go into nested regions and
   // hoist operations from there. These regions might have semantics unknown
   // to this rewriting. If the nested regions are loops, they will have been
   // processed.
+  SmallVector<LICMCandidate> LICMPotentialCandidates;
   SmallPtrSet<Operation *, 8> willBeMoved;
   for (Block &block : loop.getLoopBody()) {
     for (Operation &op : block.without_terminator()) {
       LICMCandidate candidate(op);
       if (!canBeHoisted(candidate, loop, willBeMoved, aliasAnalysis, domInfo))
         continue;
-      LICMCandidates.push_back(candidate);
+      LICMPotentialCandidates.push_back(candidate);
       willBeMoved.insert(&op);
     }
   }
+
+  // Some operations in LICMPotentialCandidates are not hoisted if the required
+  // condition is not versioned. The operations that have them as prerequisites
+  // also cannot be hoisted.
```

Comment:
Suggest: Some candidate operations require the loop to be versioned to be hoisted, if the loop is not versioned because of heuristic considerations (e.g. exceeded the versioning limit) we have to filter out the operations that depend upon them.